### PR TITLE
feat: monad stacks

### PR DESCRIPTION
**Pull request by hargoniX**
I left a few TODOs but the general concept is definitely captured I think.

### CODE REVIEW

File: lean/monad-stacks.lean [hunk 1]
```diff
@@ -1,33 +1,250 @@
...
+
+abbrev MyM (α : Type) := Reader Context (StateM State α)
+
+end Playground1
+/-!
+And this will indeed type check, however we have a very annoying issue:
+-/
+namespace Playground1
+
+#synth Monad MyM -- no monad instance :(
```
Comment (arthurpaulino) [inline lean/monad-stacks.lean:30]: What is this command for?

File: lean/monad-stacks.lean [hunk 2]
```diff
@@ -1,33 +1,250 @@
...
+
+end Playground1
+/-!
+We can fix this by manually declaring the instance and maybe even write functions
```
Comment (arthurpaulino) [inline lean/monad-stacks.lean:34]: Can you do this manually once? It's hard to grasp the idea at this point

File: lean/monad-stacks.lean [hunk 3]
```diff
@@ -1,33 +1,250 @@
...
+-/
+namespace Playground2
+abbrev Context := Nat
+abbrev State := String
+
+-- `StateM` is defined as `StateM σ α = StateT σ Id α` so the following two are equivalent.
+abbrev MyM  := ReaderT Context (StateT State Id)
+abbrev MyM2  := ReaderT Context (StateM State)
+-- This definition makes perfect sense since the `Id` monad has no effect,
+-- can you guess the definition of `Reader`?
```
Comment (arthurpaulino) [inline lean/monad-stacks.lean:51]: I don't see the reference to `Reader`

File: lean/monad-stacks.lean [hunk 4]
```diff
@@ -1,33 +1,250 @@
...
+
+end Playground2
+/-!
+So that's the monad instance for free. An other thing we would want is the
+ability to still easily `read`, `get` and `set`, after all this is the effects
+we just combined, let's see whether our monad transformers allow this:
+-/
+namespace Playground2
+
+def test : MyM String := do
```
Comment (arthurpaulino) [inline lean/monad-stacks.lean:64]: I feel like this test is doing two things at once. It's a bit hard to follow the idea at this point

File: lean/monad-stacks.lean [hunk 5]
```diff
@@ -1,33 +1,250 @@
...
+  alarm "hi"
+  pure "test"
+
+def MyM.run (context : Context) (state : State) (x : MyM α) : Sum α Error :=
+  let res := EStateM.run (ReaderT.run x context) state
+  match res with
+  | .ok v _ => Sum.inl v
+  | .error e _ => Sum.inr e
+
+#eval MyM.run 12 13 main -- Sum.inr "Alarm!: hi"
```
Comment (arthurpaulino) [inline lean/monad-stacks.lean:157]: Again, I don't understand what happened. Is it because it didn't print "test". Maybe we should be more explicit

**Review by arthurpaulino — CHANGES_REQUESTED** [review]
These are my comments for now. I will continue once the questions are addressed (I am using myself as a test case for didactics since I myself don't know this matter)

File: lean/monad-stacks.lean (hunk 1 above)
Comment (hargoniX) [reply lean/monad-stacks.lean:30]: `#synth` attempts to *synthesize* an instance of the type it is given and fails if it cannot do that.

Comment (arthurpaulino) [reply lean/monad-stacks.lean:30]: What happens if a monad instance cannot be synthesized?
